include/bitcoin/system/wallet/witness_address.hpp

```diff
@@ -0,0 +1,169 @@
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef LIBBITCOIN_SYSTEM_WALLET_WITNESS_ADDRESS_HPP
+#define LIBBITCOIN_SYSTEM_WALLET_WITNESS_ADDRESS_HPP
+
+#include <bitcoin/system/math/hash.hpp>
+#include <bitcoin/system/wallet/payment_address.hpp>
+
+namespace libbitcoin {
+namespace system {
+namespace wallet {
+
+static BC_CONSTEXPR uint8_t padding = 0x00;
+static BC_CONSTEXPR size_t witness_p2wpkh_size = 3u + short_hash_size + checksum_size;
+static BC_CONSTEXPR size_t witness_p2wsh_size = 3u + hash_size + checksum_size;
+typedef byte_array<witness_p2wpkh_size> witness_p2wpkh;
+typedef byte_array<witness_p2wsh_size> witness_p2wsh;
+
+/// A class for working with standard witness payment addresses.
+class BC_API witness_address
+  : public payment_address
+{
+public:
+    enum class encoding: uint8_t
+    {
+        // BIP 142 p2wpkh and p2wsh address versions.
+        // https://github.com/bitcoin/bips/blob/master/bip-0142.mediawiki#specification
+        mainnet_base58_p2wpkh = 0x06,
+        testnet_base58_p2wpkh = 0x03,
+        mainnet_base58_p2wsh = 0x0a,
+        testnet_base58_p2wsh = 0x28,
+
+        // These encoding values are not from a specification.
+        mainnet_p2wpkh,
+        mainnet_p2wsh,
+        testnet_p2wpkh,
+        testnet_p2wsh,
+
+        unknown
+    };
+
+    static const std::string mainnet_prefix;
+    static const std::string testnet_prefix;
+
+    typedef std::vector<witness_address> list;
+    typedef std::shared_ptr<witness_address> ptr;
+
+    /// Create base58 witness programs using provided information.
+    static void to_witness(witness_p2wpkh& out, encoding out_type,
+        uint8_t witness_version, short_hash hash);
+    static void to_witness(witness_p2wsh& out, encoding out_type,
+        uint8_t witness_version, hash_digest hash);
+
+    /// Constructors.
+    witness_address();
+    witness_address(witness_address&& other);
+    witness_address(const witness_address& other);
+    witness_address(const witness_p2wpkh& decoded,
+        encoding out_type=encoding::unknown);
+    witness_address(const witness_p2wsh& decoded,
+        encoding out_type=encoding::unknown);
+    witness_address(const data_chunk& data,
+        encoding out_type=encoding::unknown);
+    witness_address(const std::string& address,
+        encoding out_type=encoding::unknown);
+    witness_address(short_hash&& hash, encoding out_type=encoding::unknown,
```

Comment:
Avoid using the `out_` prefix for an in parameter. Just name the parameter `version` as it matches your `enum` comment and `payment_address`.